### PR TITLE
Improving speed by adding `string length match` before `regular expression match`

### PR DESCRIPTION
**Pull request by fukusuket**
## What Changed
Added `string length match` before `regular expression match`(regex matching will be skipped if the length is not same).
- Because `regular expression match` on unmatched strings is especially slow.
- But `case-insensitive match` is the default behavior in Sigma, so `exact string match` cannot be used.
  - Therefore, this PR changes it to use `string length match` instead of `exact string match`.
  - (For fields with pipes or wildcards, only `regular expression match` is applied as before)

### Environment
- OS: Windows 10 Home edition
- Hard: Memory 16GB , Core 8, SSD, laptop

### Benchmark1
I ran a benchmark using [this procedure(6.1GB evtx)](https://github.com/Yamato-Security/hayabusa/issues/778#issuecomment-1296504766) and the results were as follows.

|Version| Elapsed time | Memory(peak)  | Events with hits / Total events | Output file size(bytes)|
|-------| ------------ | --------|--------|--------|
| [before](https://github.com/Yamato-Security/hayabusa/commit/31cd18e0bf9d13e67e9cf3a262d80af39bb980b2)  | 00:13:27.170| 5.0 GiB| 1,593,715 / 4,817,181| 575085389 | 
| [This PR](https://github.com/Yamato-Security/hayabusa/commit/f13f0bb968c934fa00d07aad030abe41d1297063) | 00:12:08.121| 5.0 GiB|1,593,715 / 4,817,181 | 575085389 | 

#### Console output
[before](https://github.com/Yamato-Security/hayabusa/commit/31cd18e0bf9d13e67e9cf3a262d80af39bb980b2)
```
PS C:\tmp\hayabusa-2.1.0-win-64-bit> .\hayabusa.exe csv-timeline -d ..\hayabusa-big-evtx\ -o 1.csv --debug
...
Results Summary:

Events with hits / Total events: 1,593,715 / 4,817,181 (Data reduction: 3,223,466 events (66.92%))

Total | Unique detections: 1,627,284 | 150
Total | Unique critical detections: 0 (0.00%) | 0 (0.00%)
Total | Unique high detections: 12,044 (0.74%) | 20 (13.33%)
Total | Unique medium detections: 11,118 (0.68%) | 38 (25.33%)
Total | Unique low detections: 1,053,623 (64.75%) | 42 (28.00%)
Total | Unique informational detections: 550,499 (33.83%) | 50 (33.33%)
...
Saved file: 1.csv (575.1 MB)
Elapsed time: 00:13:27.170
Rule Parse Processing Time: 00:00:20.221
Analysis Processing Time: 00:12:43.367
Output Processing Time: 00:00:23.581

Memory usage stats:
heap stats:    peak      total      freed    current       unit      count
  reserved:    5.0 GiB    5.0 GiB   83.0 MiB    4.9 GiB
 committed:    4.6 GiB   56.5 GiB   52.0 GiB    4.5 GiB
```

[This PR](https://github.com/Yamato-Security/hayabusa/commit/f13f0bb968c934fa00d07aad030abe41d1297063)
```
PS C:\tmp\hayabusa-2.1.0-win-64-bit> .\hayabusa.exe csv-timeline -d ..\hayabusa-big-evtx\ -o 1.csv --debug
...
Results Summary:

Events with hits / Total events: 1,593,715 / 4,817,181 (Data reduction: 3,223,466 events (66.92%))

Total | Unique detections: 1,627,284 | 150
Total | Unique critical detections: 0 (0.00%) | 0 (0.00%)
Total | Unique high detections: 12,044 (0.74%) | 20 (13.33%)
Total | Unique medium detections: 11,118 (0.68%) | 38 (25.33%)
Total | Unique low detections: 1,053,623 (64.75%) | 42 (28.00%)
Total | Unique informational detections: 550,499 (33.83%) | 50 (33.33%)
...
Saved file: 1.csv (575.1 MB)
Elapsed time: 00:12:08.121
Rule Parse Processing Time: 00:00:20.157
Analysis Processing Time: 00:11:26.545
Output Processing Time: 00:00:21.417

Memory usage stats:
heap stats:    peak      total      freed    current       unit      count
  reserved:    5.0 GiB    5.0 GiB   83.0 MiB    4.9 GiB
 committed:    4.6 GiB   56.7 GiB   52.2 GiB    4.5 GiB
```

### Benchmark2
I ran a benchmark using [hayabusa-sample-evtx](https://github.com/Yamato-Security/hayabusa-sample-evtx) and the results were as follows.

|Version| Elapsed time | Memory(peak)  | Events with hits / Total events | Output file size(bytes)|
|-------| ------------ | --------|--------|--------|
| [before](https://github.com/Yamato-Security/hayabusa/commit/31cd18e0bf9d13e67e9cf3a262d80af39bb980b2)  | 00:00:17.496| 2.1 GiB| 19,606 / 47,458| 16448088| 
| [This PR](https://github.com/Yamato-Security/hayabusa/commit/f13f0bb968c934fa00d07aad030abe41d1297063) | 00:00:17.135| 2.1 GiB|19,606 / 47,458 | 16448088| 

I would appreciate it if you could review🙏

### CODE REVIEW

**Comment by hitenkoku** [comment]
@fukusuket thank you for your PR.
I have some suggestions. I will review so. please wait

File: src/detections/rule/matchers.rs [hunk 1]
```diff
@@ -274,7 +276,15 @@ impl LeafMatcher for DefaultMatcher {
...
-        if self.pipes.len() >= 2 {
+        let n = self.pipes.len();
+        if n == 0 {
+            self.match_str_len = match select_value.as_str() {
+                None => None, // strに変換できない場合は、文字列長マッチによる比較はしない
+                Some(s) if s.contains('*') | s.contains('?') => None, //ワイルドカードを含む場合は、文字列長マッチによる比較はしない
+                Some(s) => Some(s.len()),
+            };
+        }
+        if n >= 2 {
```
Comment (hitenkoku) [inline src/detections/rule/matchers.rs:287]: You can pass the extra condition check when n=0 by using "else if".

``` suggestion
    else if n >= 2 {
```

**Review by hitenkoku — CHANGES_REQUESTED** [review]
@fukusuket Thank you very much for your patience.
I have sent one suggestion.
I would appreciate it if you could check it out.

File: src/detections/rule/matchers.rs (hunk 1 above)
Comment (fukusuket) [reply src/detections/rule/matchers.rs]: Thanks for suggesting! I fixed it !

**Review by hitenkoku — APPROVED** [review]
Thank you for your fixing.
LGTM

**Review by YamatoSecurity — APPROVED** [review]
I verified that it detects all of the same alerts.
Benchmark against 14 GB data:
current main branch: 32 minutes (peak memory 7.9GB)
PR: 27 minutes 53 seconds (peak memory 7.7 GB)

13.7% speed increase! Great performance optimization!

**Comment by fukusuket** [comment]
Thank you for your review :)

**Pull request merged**